### ISSUE TITLE
Wrong report branches_sharing_code

### ISSUE THREAD

**Issue by douchuan** [rust-lang/rust-clippy#7589]
Lint name:
#[warn(clippy::branches_sharing_code)]

I tried this code:

```rust
pub unsafe fn insert<K, V>(
    root: Option<NonNull<Node<K, V>>>,
    key: K,
    val: V,
) -> Result<NonNull<Node<K, V>>, ()>
where
    K: Ord,
{
    let mut nq = NodeQuery::new(root);
    let mut parent = None;
    while nq.is_some() {
        parent = nq.node;
        match key.cmp(nq.get_key().unwrap()) {
            Ordering::Less => nq = nq.left(),
            Ordering::Greater => nq = nq.right(),
            Ordering::Equal => {
                nq.set_entry((key, Some(val))); // update val
                return Err(());
            }
        }
    }

    //插入x
    let mut x;
    if let Some(mut node) = parent {
        if key < node.as_ref().key {
            x = Node::new_entry(key, val);
            node.as_mut().left = Some(x);
        } else {
            x = Node::new_entry(key, val);
            node.as_mut().right = Some(x);
        }
        x.as_mut().parent = parent;
    } else {
        x = Node::new_entry(key, val);
    }

    Ok(x)
}
```

I expected to see this happen:  **key** is not **Copy**, so should not report the warning.

Instead, this happened: **key** is not **Copy**, clippy advice extract common codes

```
warning: all if blocks contain the same code at the start
  --> src/tree/binary/bst.rs:92:9
   |
92 | /         if key < node.as_ref().key {
93 | |             x = Node::new_entry(key, val);
   | |__________________________________________^
   |
   = note: `#[warn(clippy::branches_sharing_code)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#branches_sharing_code
help: consider moving the start statements out like this
   |
92 |         x = Node::new_entry(key, val);
93 |         if key < node.as_ref().key {

```

### Meta

- `cargo clippy -V`: clippy 0.1.53 (53cb7b09 2021-06-17)
- `rustc -Vv`:
  ```
rustc 1.53.0 (53cb7b09b 2021-06-17)
binary: rustc
commit-hash: 53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b
commit-date: 2021-06-17
host: x86_64-apple-darwin
release: 1.53.0
LLVM version: 12.0.1
  ```

**Comment by BearOve** [comment]
I noticed the same issue today in a slightly different scenario. My gut tells me the root cause is the same. If I follow the suggestion the code no longer compiles.

```rust
struct SomeStruct {
    complex_member: String,
    basic_member: bool,
}

struct BaseStruct {
    member: Option<SomeStruct>,
}

impl BaseStruct {
    fn handle_struct(&mut self, arg: Option<SomeStruct>) -> Option<String> {
        if let Some(arg) = arg {
            if arg.basic_member {
                self.member = Some(arg);
                return self.handle_member();
            } else {
                self.member = Some(arg);
            }
        }
        None
    }

    fn handle_member(&mut self) -> Option<String> {
        Some(self.member.take()?.complex_member)
    }
}
```

**Comment by llogiq** [comment]
@BearOve I cannot reproduce any clippy warning with your code using either current stable or nightly. Can you try with a newer toolchain?

**Comment by BearOve** [comment]
NOTE: When I run with `#![deny(clippy::all)]` on stable it includes `branches_sharing_code`, when I run the same on nightly it is not included and no warning is emitted. However, if I enable the lint explicitly it does not matter if it is stable or not.

Steps used:
1. I ran cargo new clippy-bug
2. I added the code to the top of main.rs
3. I enabled the clippy lint by adding `#![deny(clippy::branches_sharing_code)]` to the top of main.rs
4. I ran cargo clippy

# Toolchain and rust version
```
[boh@boh-xps clippy-bug]$ rustup default
stable-x86_64-unknown-linux-gnu (default)
[boh@boh-xps clippy-bug]$ 
[boh@boh-xps clippy-bug]$ rustup default
stable-x86_64-unknown-linux-gnu (default)
[boh@boh-xps clippy-bug]$ cargo -V
cargo 1.56.0 (4ed5d137b 2021-10-04)
[boh@boh-xps clippy-bug]$ cargo clippy -V
clippy 0.1.56 (09c42c4 2021-10-18)
[boh@boh-xps clippy-bug]$ rustup update
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'

   stable-x86_64-unknown-linux-gnu unchanged - rustc 1.56.0 (09c42c458 2021-10-18)
  nightly-x86_64-unknown-linux-gnu unchanged - rustc 1.58.0-nightly (ff0e14829 2021-10-31)
```

# Using latest stable
```
[boh@boh-xps clippy-bug]$ cargo clippy
    Checking clippy-bug v0.1.0 (/home/boh/clippy-bug)
error: all if blocks contain the same code at the start
  --> src/main.rs:16:13
   |
16 | /             if arg.basic_member {
17 | |                 self.member = Some(arg);
   | |________________________________________^
   |
note: the lint level is defined here
  --> src/main.rs:2:9
   |
2  | #![deny(clippy::branches_sharing_code)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#branches_sharing_code
help: consider moving the start statements out like this
   |
16 ~             self.member = Some(arg);
17 +             if arg.basic_member {
   |

error: could not compile `clippy-bug` due to previous error
```

# Using latest nightly
```
boh@boh-xps clippy-bug]$ cargo +nightly clippy
    Checking clippy-bug v0.1.0 (/home/boh/clippy-bug)
error: all if blocks contain the same code at the start
  --> src/main.rs:16:13
   |
16 | /             if arg.basic_member {
17 | |                 self.member = Some(arg);
   | |________________________________________^
   |
note: the lint level is defined here
  --> src/main.rs:2:9
   |
2  | #![deny(clippy::branches_sharing_code)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#branches_sharing_code
help: consider moving the start statements out like this
   |
16 ~             self.member = Some(arg);
17 +             if arg.basic_member {
   |

error: could not compile `clippy-bug` due to previous error
```

**Comment by xFrednet** [comment]
The `branches_sharing_code` lint has been moved to the `nursery` group due to this and other false positives. The plan is to fix them and then re-enable it by default :upside_down_face:
